File: app/registry/ghcr.py
from typing import Optional

import requests

from app.config import log
import app.http as _http

_MAX_PAGES = 100  # Safety cap to prevent infinite loops (100 × 100 = 10k versions)


def _fetch_ghcr_tags(image_name: str, github_token: str, current_tag: Optional[str] = None) -> list[str]:
    """
    Fetch tags from GitHub Container Registry using the GitHub Packages REST API.
    Uses the versions endpoint which returns results sorted by creation date
    (newest first), allowing early stop once the current tag is found.
    """
    # Strip ghcr.io/ or lscr.io/ → owner/repo (may be owner/repo or owner/group/repo)
    path = image_name.removeprefix("ghcr.io/").removeprefix("lscr.io/")
    parts = path.split("/")
    if len(parts) < 2:
        log.warning(f"GHCR: cannot parse owner/repo from '{image_name}'")
        return []
    owner = parts[0]
    repo = "/".join(parts[1:])

    if not github_token:
        log.warning(f"GHCR: no GITHUB_TOKEN set — cannot fetch tags for {image_name}. "
                    "Set GITHUB_TOKEN with read:packages scope.")
        return []

    headers = {
        "Accept": "application/vnd.github+json",
        "Authorization": f"Bearer {github_token}",
        "X-GitHub-Api-Version": "2022-11-28",
    }
    tags: list[str] = []
    page = 1
    # Try org endpoint first; fall back to user endpoint on 404
    base_url = (f"https://api.github.com/orgs/{owner}/packages/container/"
                f"{repo}/versions")

    while page <= _MAX_PAGES:
        url = f"{base_url}?per_page=100&page={page}"
        try:
            resp = _http.http_session.get(url, headers=headers, timeout=20)

            # If org endpoint returns 404 on first page, switch to user endpoint
            if resp.status_code == 404 and page == 1:
                base_url = (f"https://api.github.com/users/{owner}/packages/container/"
                            f"{repo}/versions")
                url = f"{base_url}?per_page=100&page={page}"
                resp = _http.http_session.get(url, headers=headers, timeout=20)

            resp.raise_for_status()
            versions = resp.json()

            if not versions:
                break

            found_current = False
            for version in versions:
                version_tags = (version.get("metadata", {})
                                .get("container", {})
                                .get("tags") or [])
                tags.extend(version_tags)
                if current_tag and current_tag in version_tags:
                    found_current = True

            if found_current:
                log.debug(f"GHCR: found current tag '{current_tag}' — stopping pagination")
                break

            page += 1
        except requests.HTTPError as exc:
            log.error(f"GHCR HTTP error for {owner}/{repo}: {exc}")
            break
        except Exception as exc:
            log.error(f"GHCR error for {owner}/{repo}: {exc}")
            break

    return tags

File: src/registry/ghcr.rs
use reqwest::{header::HeaderMap, Client};
use serde::Deserialize;

const MAX_PAGES: u32 = 100;

#[derive(Deserialize)]
struct VersionEntry {
    metadata: Option<MetadataEntry>,
}

#[derive(Deserialize)]
struct MetadataEntry {
    container: Option<ContainerEntry>,
}

#[derive(Deserialize)]
struct ContainerEntry {
    tags: Option<Vec<String>>,
}

pub async fn fetch_ghcr_tags(
    client: &Client,
    image: &str,
    github_token: &str,
    current_tag: Option<&str>,
    api_base: &str,
) -> Vec<String> {
    if github_token.is_empty() {
        return vec![];
    }

    let path = image
        .strip_prefix("ghcr.io/")
        .or_else(|| image.strip_prefix("lscr.io/"))
        .unwrap_or(image);
    let parts: Vec<&str> = path.split('/').collect();
    if parts.len() < 2 {
        return vec![];
    }
    let owner = parts[0];
    let repo = parts[1..].join("/");

    let mut headers = HeaderMap::new();
    headers.insert("Accept", "application/vnd.github+json".parse().unwrap());
    headers.insert(
        "Authorization",
        format!("Bearer {github_token}").parse().unwrap(),
    );
    headers.insert("X-GitHub-Api-Version", "2022-11-28".parse().unwrap());

    let mut tags = Vec::new();
    let mut page = 1u32;
    let mut base_url = format!("{api_base}/orgs/{owner}/packages/container/{repo}/versions");

    while page <= MAX_PAGES {
        let url = format!("{base_url}?per_page=100&page={page}");
        let mut resp_result = client.get(&url).headers(headers.clone()).send().await;

        // On first page, if org endpoint 404s, fall back to user endpoint
        if page == 1 {
            if let Ok(ref r) = resp_result {
                if r.status() == 404 {
                    base_url =
                        format!("{api_base}/users/{owner}/packages/container/{repo}/versions");
                    let url2 = format!("{base_url}?per_page=100&page={page}");
                    resp_result = client.get(&url2).headers(headers.clone()).send().await;
                }
            }
        }

        let resp = match resp_result {
            Ok(r) if r.status().is_success() => r,
            Ok(_) => break,
            Err(_) => break,
        };

        let versions: Vec<VersionEntry> = match resp.json().await {
            Ok(v) => v,
            Err(_) => break,
        };

        if versions.is_empty() {
            break;
        }

        let mut found_current = false;
        for version in versions {
            let version_tags: Vec<String> = version
                .metadata
                .and_then(|m| m.container)
                .and_then(|c| c.tags)
                .unwrap_or_default();
            if let Some(ct) = current_tag {
                if version_tags.iter().any(|t| t == ct) {
                    found_current = true;
                }
            }
            tags.extend(version_tags);
        }

        if found_current {
            break;
        }
        page += 1;
    }

    tags
}
